File: app/src/main/java/com/commit451/gitlab/fragment/PickBranchFragment.java
package com.commit451.gitlab.fragment;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.commit451.gitlab.App;
import com.commit451.gitlab.R;
import com.commit451.gitlab.activity.PickBranchOrTagActivity;
import com.commit451.gitlab.adapter.BranchesAdapter;
import com.commit451.gitlab.model.Ref;
import com.commit451.gitlab.model.api.Branch;
import com.commit451.gitlab.rx.CustomSingleObserver;

import org.parceler.Parcels;

import java.util.List;

import butterknife.BindView;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

/**
 * Pick a branch, any branch
 */
public class PickBranchFragment extends ButterKnifeFragment {

    private static final String EXTRA_PROJECT_ID = "project_id";
    private static final String EXTRA_REF = "ref";

    public static PickBranchFragment newInstance(long projectId, @Nullable Ref ref) {
        PickBranchFragment fragment = new PickBranchFragment();
        Bundle args = new Bundle();
        args.putLong(EXTRA_PROJECT_ID, projectId);
        args.putParcelable(EXTRA_REF, Parcels.wrap(ref));
        fragment.setArguments(args);
        return fragment;
    }

    @BindView(R.id.list)
    RecyclerView mProjectsListView;
    @BindView(R.id.message_text)
    TextView mMessageView;
    @BindView(R.id.progress)
    View mProgress;

    BranchesAdapter mBranchesAdapter;

    long mProjectId;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mProjectId = getArguments().getLong(EXTRA_PROJECT_ID);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_pick_branch, container, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        Ref existingRef = Parcels.unwrap(getArguments().getParcelable(EXTRA_REF));
        mBranchesAdapter = new BranchesAdapter(existingRef, new BranchesAdapter.Listener() {
            @Override
            public void onBranchClicked(Branch entry) {
                Intent data = new Intent();
                Ref ref = new Ref(Ref.TYPE_BRANCH, entry.getName());
                data.putExtra(PickBranchOrTagActivity.EXTRA_REF, Parcels.wrap(ref));
                getActivity().setResult(Activity.RESULT_OK, data);
                getActivity().finish();
            }
        });
        mProjectsListView.setLayoutManager(new LinearLayoutManager(getActivity()));
        mProjectsListView.setAdapter(mBranchesAdapter);

        loadData();
    }

    @Override
    protected void loadData() {
        if (getView() == null) {
            return;
        }
        mProgress.setVisibility(View.VISIBLE);
        mMessageView.setVisibility(View.GONE);

        App.get().getGitLab().getBranches(mProjectId)
                .compose(this.<List<Branch>>bindToLifecycle())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new CustomSingleObserver<List<Branch>>() {

                    @Override
                    public void error(Throwable e) {
                        Timber.e(e);
                        mProgress.setVisibility(View.GONE);
                        mMessageView.setVisibility(View.VISIBLE);
                    }

                    @Override
                    public void success(List<Branch> branches) {
                        mProgress.setVisibility(View.GONE);
                        mBranchesAdapter.setEntries(branches);
                    }
                });
    }

}

File: app/src/main/java/com/commit451/gitlab/fragment/PickBranchFragment.kt
package com.commit451.gitlab.fragment

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.LinearLayoutManager
import com.commit451.gitlab.App
import com.commit451.gitlab.R
import com.commit451.gitlab.activity.PickBranchOrTagActivity
import com.commit451.gitlab.adapter.BranchAdapter
import com.commit451.gitlab.extension.mapResponseSuccessWithPaginationData
import com.commit451.gitlab.extension.with
import com.commit451.gitlab.model.Ref
import com.commit451.gitlab.model.api.Branch
import com.commit451.gitlab.util.OnScrollLoadMoreListener
import kotlinx.android.synthetic.main.fragment_pick_branch.*
import kotlinx.android.synthetic.main.progress.*
import timber.log.Timber

/**
 * Pick a branch, any branch
 */
class PickBranchFragment : BaseFragment() {

    companion object {

        private const val EXTRA_PROJECT_ID = "project_id"
        private const val EXTRA_REF = "ref"

        fun newInstance(projectId: Long, ref: Ref?): PickBranchFragment {
            val fragment = PickBranchFragment()
            val args = Bundle()
            args.putLong(EXTRA_PROJECT_ID, projectId)
            args.putParcelable(EXTRA_REF, ref)
            fragment.arguments = args
            return fragment
        }
    }

    private lateinit var adapterBranches: BranchAdapter

    private var projectId: Long = 0

    private var nextPageUrl: String? = null
    private var loading: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        projectId = arguments?.getLong(EXTRA_PROJECT_ID)!!
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_pick_branch, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val existingRef = arguments?.getParcelable<Ref>(EXTRA_REF)!!
        adapterBranches = BranchAdapter(existingRef, object : BranchAdapter.Listener {
            override fun onBranchClicked(entry: Branch) {
                val data = Intent()
                val ref = Ref(Ref.TYPE_BRANCH, entry.name)
                data.putExtra(PickBranchOrTagActivity.EXTRA_REF, ref)
                activity?.setResult(Activity.RESULT_OK, data)
                activity?.finish()
            }
        })
        val layoutManager = LinearLayoutManager(activity)
        listProjects.layoutManager = layoutManager
        listProjects.adapter = adapterBranches
        listProjects.addOnScrollListener(OnScrollLoadMoreListener(layoutManager, {
            !loading && nextPageUrl != null
        }, {
            loadMore()
        }))

        loadData()
    }

    override fun loadData() {
        loading = true
        progress.visibility = View.VISIBLE
        textMessage.visibility = View.GONE

        App.get().gitLab.getBranches(projectId)
                .mapResponseSuccessWithPaginationData()
                .with(this)
                .subscribe({
                    loading = false
                    nextPageUrl = it.paginationData.next
                    progress.visibility = View.GONE
                    adapterBranches.setEntries(it.body)
                }, {
                    Timber.e(it)
                    progress.visibility = View.GONE
                    textMessage.visibility = View.VISIBLE
                })
    }

    fun loadMore() {
        loading = true
        App.get().gitLab.getBranches(nextPageUrl.toString())
                .mapResponseSuccessWithPaginationData()
                .with(this)
                .subscribe({
                    loading = false
                    nextPageUrl = it.paginationData.next
                    adapterBranches.addEntries(it.body)
                }, {
                    Timber.e(it)
                    loading = false
                })
    }
}
